src/util/lazy.h

```diff
@@ -0,0 +1,46 @@
+/*******************************************************************\
+
+Module: Util
+
+Author: Romain Brenguier, romain.brenguier@diffblue.com
+
+\*******************************************************************/
+
+#ifndef CPROVER_UTIL_LAZY_H
+#define CPROVER_UTIL_LAZY_H
+
+#include <functional>
+#include <util/optional.h>
+
+template <typename valuet, typename infot>
+class lazyt
+{
+public:
+  /// Delay the computation of \p fun to the next time the \c force method
+  /// is called.
+  static lazyt from_fun(std::function<valuet(infot)> fun)
+  {
+    return lazyt{std::move(fun)};
+  }
+
+  /// Force the computation of the value. If it was already computed (even if
+  /// \p info was different), return the same result.
+  valuet force(infot info)
```

Comment:
I find "force" an unintuitive name. How about "evaluate", "compute" or "force_evaluation"?

Reply:
`force` is used by .net, scheme and ocaml https://en.wikipedia.org/wiki/Lazy_evaluation so I think it's relatively standard, but I don't know if someone encountered similar concepts in other libraries or languages?

Reply:
If it's a standard term then that's fine.